File: client/js/protocol.ts
export interface IceCandidate {
	candidate: string
	sdpMid: string | null
	sdpMLineIndex: number | null
	usernameFragment: string | null
}

export class SignalMessage {
	SessionJoin() {
		return {
			Encode: (sessionId: string, isHost: boolean) => JSON.stringify({ SessionJoin: [sessionId, isHost] }),
		}
	}

	SessionReady() {
		return {
			Encode: (sessionId: string, userId: number) => JSON.stringify({ SessionReady: [sessionId, userId] }),
			Decode: (data: { SessionReady: any[] }): { sessionId: string; userId: number } => {
				return {
					sessionId: data.SessionReady[0],
					userId: data.SessionReady[1],
				}
			},
		}
	}

	SdpOffer() {
		return {
			Encode: (sessionId: string, userId: number, offer: string) => JSON.stringify({ SdpOffer: [sessionId, userId, offer] }),
			Decode: (data: { SdpOffer: any[] }): { sessionId: string; userId: number; offer: string } => {
				return {
					sessionId: data.SdpOffer[0],
					userId: data.SdpOffer[1],
					offer: data.SdpOffer[2],
				}
			},
		}
	}

	SdpAnswer() {
		return {
			Encode: (sessionId: string, userId: number, answer: string) => JSON.stringify({ SdpAnswer: [sessionId, userId, answer] }),
			Decode: (data: { SdpAnswer: any[] }): { sessionId: string; userId: number; answer: string } => {
				return {
					sessionId: data.SdpAnswer[0],
					userId: data.SdpAnswer[1],
					answer: data.SdpAnswer[2],
				}
			},
		}
	}

	IceCandidate() {
		return {
			Encode: (sessionId: string, userId: number, candidate: IceCandidate) =>
				JSON.stringify({ IceCandidate: [sessionId, userId, JSON.stringify(candidate)] }),
			Decode: (data: { IceCandidate: any[] }): { sessionId: string; userId: number; candidate: IceCandidate } => {
				return {
					sessionId: data.IceCandidate[0],
					userId: data.IceCandidate[1],
					candidate: JSON.parse(data.IceCandidate[2]),
				}
			},
		}
	}
}

File: client/cs/Protocol.cs
using System.Text.Json;

namespace ezrtc
{
	public class ICandidate
	{
		public string candidate { get; set; }
		public string sdpMid { get; set; }
		public int sdpMLineIndex { get; set; }
		public string usernameFragment { get; set; }
	}

	public class Status
	{
		public string? session_id { get; set; }
		public bool? is_host { get; set; }
		public string? version { get; set; }
	}

	public class SignalMessage
	{
		public class KeepAlive
		{
			public class KeepAliveInput
			{
				public object[] KeepAlive { get; set; }
			}

			public class KeepAliveOutput
			{
				public string userId;
				public Status status;
			}

			public static KeepAliveOutput Decode(string data)
			{
				var message = JsonSerializer.Deserialize<KeepAliveInput>(data);
				object[] keepAlive = message.KeepAlive;

				return new KeepAliveOutput
				{
					userId = keepAlive[0].ToString(),
					status = JsonSerializer.Deserialize<Status>(keepAlive[1].ToString())
				};
			}

			public static string Encode(string userId, Status status)
			{
				var message = new { KeepAlive = new object[] { Convert.ToUInt32(userId), status } };
				return JsonSerializer.Serialize(message);
			}
		}

		public class SessionJoin
		{
			public static string Encode(string sessionId, bool isHost)
			{
				var message = new { SessionJoin = new object[] { sessionId, isHost } };
				return JsonSerializer.Serialize(message);
			}
		}

		public class SessionReady
		{
			public class SessionReadyInput
			{
				public object[] SessionReady { get; set; }
			}

			public class SessionReadyOutput
			{
				public string sessionId;
				public string userId;
			}

			public static SessionReadyOutput Decode(string data)
			{
				var message = JsonSerializer.Deserialize<SessionReadyInput>(data);
				object[] sessionReady = message.SessionReady;

				return new SessionReadyOutput
				{
					sessionId = sessionReady[0].ToString(),
					userId = sessionReady[1].ToString(),
				};
			}
		}

		public class SdpOffer
		{
			public class SdpOfferInput
			{
				public object[] SdpOffer { get; set; }
			}

			public class SdpOfferOutput
			{
				public string sessionId;
				public string userId;
				public string offer;
			}

			public static string Encode(string sessionId, string userId, string offer)
			{
				var message = new { SdpOffer = new object[] { sessionId, Convert.ToUInt32(userId), offer } };
				return JsonSerializer.Serialize(message);
			}

			public static SdpOfferOutput Decode(string data)
			{
				var message = JsonSerializer.Deserialize<SdpOfferInput>(data);
				object[] sdpOffer = message.SdpOffer;

				return new SdpOfferOutput
				{
					sessionId = sdpOffer[0].ToString(),
					userId = sdpOffer[1].ToString(),
					offer = sdpOffer[2].ToString()
				};
			}
		}

		public class SdpAnswer
		{
			public class SdpAnswerInput
			{
				public object[] SdpAnswer { get; set; }
			}

			public class SdpAnswerOutput
			{
				public string sessionId;
				public string userId;
				public string answer;
			}

			public static SdpAnswerOutput Decode(string data)
			{
				var message = JsonSerializer.Deserialize<SdpAnswerInput>(data);
				object[] sdpOffer = message.SdpAnswer;

				return new SdpAnswerOutput
				{
					sessionId = sdpOffer[0].ToString(),
					userId = sdpOffer[1].ToString(),
					answer = sdpOffer[2].ToString()
				};
			}

			public static string Encode(string sessionId, string userId, string answer)
			{
				var message = new { SdpAnswer = new object[] { sessionId, Convert.ToUInt32(userId), answer } };
				return JsonSerializer.Serialize(message);
			}
		}

		public class IceCandidate
		{
			public class IceCandidateInput
			{
				public object[] IceCandidate { get; set; }
			}

			public class IceCandidateOutput
			{
				public string sessionId;
				public string userId;
				public ICandidate candidate;
			}

			public static IceCandidateOutput Decode(string data)
			{
				var message = JsonSerializer.Deserialize<IceCandidateInput>(data);
				object[] sdpOffer = message.IceCandidate;

				return new IceCandidateOutput
				{
					sessionId = sdpOffer[0].ToString(),
					userId = sdpOffer[1].ToString(),
					candidate = JsonSerializer.Deserialize<ICandidate>(sdpOffer[2].ToString()),
				};
			}
		}
	}
}
